### PR TITLE
Communication

### PR DESCRIPTION
**Pull request by sanmit**

### CODE REVIEW

File: bin/HFO [hunk 1]
```diff
@@ -138,6 +140,8 @@ def parseArgs():
                  help='Server provides full-state information to agents.')
   p.add_argument('--seed', dest='seed', type=int, default=-1,
                  help='Seed the server\'s RNG. Default: time.')
+  p.add_argument('--messageSize', dest='messageSize', type=int, default=1000,
```
Comment (mhauskn) [inline bin/HFO]: Please change this to '--message-size' in order to conform with style of the other arguments.

File: example/mid_level_move_agent.cpp [hunk 2]
```diff
@@ -31,12 +31,15 @@ int main(int argc, char** argv) {
     }
     std::cout << "target (x,y) = " << target_x << ", " << target_y << std::endl;
     while (status == IN_GAME) {
+
```
Comment (mhauskn) [inline example/mid_level_move_agent.cpp]: Please remove blank lines.

File: example/hfo_example_agent.py [hunk 3]
```diff
@@ -1,9 +1,16 @@
 #!/usr/bin/env python
 # encoding: utf-8
 
+import sys
+
 # First Start the server: $> bin/start.py
 
 if __name__ == '__main__':
+
```
Comment (mhauskn) [inline example/hfo_example_agent.py]: Please remove blank lines.

File: example/mid_level_move_agent.cpp [hunk 4]
```diff
@@ -31,12 +31,15 @@ int main(int argc, char** argv) {
...
       // Get the vector of state features for the current state
       const vector<float>& feature_vec = hfo.getState();
       // Perform the action
       hfo.act(MOVE_TO, target_x, target_y);
       // Advance the environment and get the game status
       status = hfo.step();
+    
+      std::cout << "Status: " << status << " (IN_GAME = " << IN_GAME << ")\n";
```
Comment (mhauskn) [inline example/mid_level_move_agent.cpp]: What is the purpose of this print?

File: src/HFO.cpp [hunk 5]
```diff
@@ -188,6 +188,28 @@ void HFOEnvironment::connectToAgentServer(int server_port,
...
+    std::vector<char> hearMsgBuffer;
+    hearMsgBuffer.resize(msgLength);
+    if (recv(sockfd, &hearMsgBuffer[0], msgLength, 0) < 0){
+      perror("[Agent Client] ERROR recieving hear message from socket");
+      close(sockfd);
+      exit(1);
+    }  
+    hear_msg.assign(&(hearMsgBuffer[0]), hearMsgBuffer.size());
+  }
+
```
Comment (mhauskn) [inline src/HFO.cpp]: Remove blank lines.

File: src/agent.cpp [hunk 6]
```diff
@@ -120,8 +122,10 @@ Agent::Agent()
       M_field_evaluator(createFieldEvaluator()),
       M_action_generator(createActionGenerator()),
       lastTrainerMessageTime(-1),
+      lastTeammateMessageTime(-1),
       server_port(6008),
       client_connected(false),
+
```
Comment (mhauskn) [inline src/agent.cpp]: Remove blank line.

File: src/agent.cpp [hunk 7]
```diff
@@ -443,8 +477,44 @@ void Agent::actionImpl() {
...
+  // [Sanmit] Receive the outgoing communication
+  // Receive message length
+  uint32_t sayMsgLength;
+  if (recv(newsockfd, &sayMsgLength, sizeof(uint32_t), 0) < 0){
+    perror("[Agent Server] ERROR recv size of say message from socket");
+    close(sockfd);
+    exit(1);
+  }
+
+  std::cout << "[AGENT.CPP] Say message length " << sayMsgLength << std::endl;
```
Comment (mhauskn) [inline src/agent.cpp]: Remove this line.

File: src/agent.cpp [hunk 8]
```diff
@@ -443,8 +477,44 @@ void Agent::actionImpl() {
...
+  std::string msgString = "";
+
+  // Check message size
+  if (sayMsgLength > ServerParam::i().playerSayMsgSize()){
+    perror("[Agent Server] ERROR message size too large. Increase size by starting bin/HFO with larger --messageSize argument");
+    close(sockfd);
+    exit(1);
+  }
+  if (sayMsgLength > 0) { 
+
```
Comment (mhauskn) [inline src/agent.cpp]: Remove blank line.

File: src/agent.cpp [hunk 9]
```diff
@@ -704,7 +774,8 @@ Agent::communicationImpl()
 {
     if ( M_communication )
     {
-        M_communication->execute( this );
+      // [Sanmit]: Turning this off since it adds default communication messages which can conflict with our comm messages.
```
Comment (mhauskn) [inline src/agent.cpp]: Do we have any idea of how much impact this has on the Agent2D policies?

**Comment by mhauskn** [comment]
PR looks good - I made some minor comments about removing various blank lines and prints.

File: src/agent.cpp (hunk 9 above)
Comment (sanmit) [inline src/agent.cpp]: By default, the communication policy of Agent2D is round robin (i.e. players take turns who communicates). They communicate the location of the ball, players, and stamina. I'd expect that most of this information is redundant with full state turned on, but the only way to know for sure would be to test it.

**Pull request merged**